### PR TITLE
Fix scan carry types in gradient of complex ODE

### PR DESCRIPTION
**Pull request by dav0dea**
This pull request fixes issue #3986. The included test case is currently **not** working properly, I am creating this pull request for further discussion.

> To make it slightly more comprehensive, I might suggest also adding a single complex-valued parameter `alpha` with which to differentiate against as well, i.e., testing the ODE `∂y/∂t = α y`.

As @shoyer suggested, I have added a complex-valued `alpha` in the differential equation but this causes the gradient to be wrong:
```python
def dy_dt(vals, t):
  alpha, y = vals
  return alpha, alpha * y
```
The assertion error from `check_grads()` is:
```python
AssertionError:
Not equal to tolerance rtol=0.1, atol=0.1

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 375.89012743
Max relative difference: 0.53135127
x: array(-331.532907)
y: array(-707.42303, dtype=float32)
```
If `alpha` is real or not present in the differential equation the test works and also correctly checks for the problem that was fixed in the first place.

Interestingly enough, everything also works correctly if `alpha` is added to `y`, subtracted from `y`, divided by `y` or dividing `y`. The gradient mismatch only occurs if `alpha` is multiplied with `y`. I am not sure why this might be the case, could there be another bug causing this behavior?

### CODE REVIEW

File: tests/ode_test.py [hunk 1]
```diff
@@ -232,6 +232,22 @@ def f(z):
 
     jax.grad(f)(jnp.ones(2))  # doesn't crash
 
+  @jtu.skip_on_devices("tpu", "gpu")
+  def test_complex_odeint(self):
+    # https://github.com/google/jax/issues/3986
+
+    def dy_dt(vals, t):
+      alpha, y = vals
+      return alpha, alpha * y
```
Comment (shoyer) [inline tests/ode_test.py:241]: Maybe try making `alpha` a "parameter" here rather than something that is explicitly integrated?
```python
def dy_dt(y, t, alpha):
  return alpha * y
```

**Comment by shoyer** [comment]
> Interestingly enough, everything also works correctly if `alpha` is added to `y`, subtracted from `y`, divided by `y` or dividing `y`. The gradient mismatch only occurs if `alpha` is multiplied with `y`. I am not sure why this might be the case, could there be another bug causing this behavior?

Indeed, sounds like this _might_ be another bug.

The mathematical condition for a well-defined gradient for a complex-to-complex function is that the function is holomorphic: https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Complex-numbers-and-differentiation

I don't know the solution of an ODE satisfies that condition.

I would consider testing a wrapper function around `odeint` rather than testing `integrate` directly, and having that wrapping function returning something like the real-part of the solution instead. That way we could at least rule out the "holomorphic" part when considering whether or not this is a bug.

File: tests/ode_test.py (hunk 1 above)
Comment (PhilippThoelke) [reply tests/ode_test.py]: I have added this in the newest commit but it still produces the same error.

**Comment by PhilippThoelke** [comment]
I have now wrapped the `odeint` function to be a complex-to-real valued function so we do not need the holomorphic condition. Unfortunately, the gradients still don't match the numerical ones. Changing which parameters are passed to the wrapper function `f` (i.e. just `y0` or `y0` and `alpha` etc.) also has no effect on the error.

**Comment by shoyer** [comment]
I think it should be working as of the latest commit. It looks like estimated `error_ratio` also had to be fixed up to always be a real valued number.

**Comment by PhilippThoelke** [comment]
Okay I see, thanks for helping out!

**Comment by shoyer** [comment]
@PhilippThoelke thanks for your contribution here! I think this is definitely an improvement over the current state of affairs, so let's give it a try and see if any issues come up.
